MuStop Generator

AndrewEdmonds11 opened the pull request:
Hi Everyone

This is a draft pull request because I would appreciate any input people have on this new MuStopGenerator. In particular, if there is code or configuration that is unclear, or if anything should be implemented differently.

Thanks in advance for any help,
Andy

FNALbuild commented:
Hi @AndrewEdmonds11,
You have proposed changes to files in these packages:
- EventGenerator
- JobConfig
- Mu2eUtilities
- Mu2eG4
- GlobalConstantsService

which require these tests: build.




<a href="https://mu2ewiki.fnal.gov/wiki/Git#GitHub_Pull_Request_Procedures_and_FNALbuild">FNALbuild is explained here.</a>

kutschke commented:
@FNALbuild run build tests

FNALbuild commented:
:hourglass: The following tests have been triggered for ref https://github.com/Mu2e/Offline/pull/231/commits/3443e1dec64e774c8a4e9762f9c53844fbe6a09c: build

FNALbuild commented:
:sunny: The tests passed at ref 3443e1dec64e774c8a4e9762f9c53844fbe6a09c.


| Test          | Result        | Details |
| ------------- |:-------------:| ------- |
| merge | :heavy_check_mark: | Merged 3443e1dec64e774c8a4e9762f9c53844fbe6a09c at a95b761eb22a55040055b0cd1a82b5b1dbea5e6d |
| scons build (prof) | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//177/artifact/scons.log). Build time: 13 min 34 sec |
| ceSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//177/artifact/ceSimReco.log) |
| g4test_03MT | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//177/artifact/g4test_03MT.log) |
| transportOnly | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//177/artifact/transportOnly.log) |
| PS | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//177/artifact/PS.log) |
| g4study | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//177/artifact/g4study.log) |
| cosmicSimReco | :heavy_check_mark: | [Log file](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//177/artifact/cosmicSimReco.log) |
| FIXME, TODO count | :heavy_check_mark: | [TODO (0) FIXME (0) in 6 files](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//177/artifact/fixme_todo.log) |
| clang-tidy | :wavy_dash: | [0 errors 3 warnings](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//177/artifact/clang-tidy.log) |

For more information, please check the job page [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-build-test//177/console).

brownd1978 commented:
Can you please explain the normalization model of this module?  Does each event corresponds to the decay or capture of a single muon, with the output containing all its products?  If so, is there any attempt to make sure energy and momentum and other quantum numbers are conserved?  Or does each event corresponds to an incoherent sampling of independent muons, which just happen to stop at the same point, each of which may or may not produce a particle from the given process?  Does each generator process always create a particle in every event?  If so, how is the relative rate of (say) photon and neutron production maintained?

Can you create a test sample, maybe also in TTree form, so that we can look at what it produces?

brownd1978 commented:
In the meeting you presented this work, the request was made to expand this module to also produce primary events (conversion electrons, ...).  Looking at the structure I don't see how that's natural, given that this module intrinsically generate multiple decay products.  I can see how some infrastructure (GenProc, ...) might be shared with a module designed to produce primary events.  That suggests those structs be split out of the module and put in their own files, and maybe made into classes.

AndrewEdmonds11 commented:
Hi Dave

> Can you please explain the normalization model of this module? Does each event corresponds to the decay or capture of a single muon, with the output containing all its products? 

Yes, each event corresponds to a single stopped muon that will either decay or capture. The user configures lists of processes that can occur for stopped muons, decaying muons, and captured muons. Each process has either a random chance of producing one particle (e.g. 79% for a 2p-1s X-ray), or a Poisson mean to produce multiple particles of the specified type (e.g. 1.2 for neutrons) 

The generator will go through the list of stopped muon processes in every event, but will only go through *either* the list of decay processes *or* the list of capture processes depending on the capture fraction.

> If so, is there any attempt to make sure energy and momentum and other quantum numbers are conserved? 

No, there isn't an attempt to conserve energy and momentum. I think the nuclear physics of the capture process complicates things too much.

> Or does each event corresponds to an incoherent sampling of independent muons, which just happen to stop at the same point, each of which may or may not produce a particle from the given process? 

I'm not sure. I don't quite understand the question, but I will think on it further.

>Does each generator process always create a particle in every event? If so, how is the relative rate of (say) photon and neutron production maintained?

No, each generator does not always produce a particle. It will depend on the emission rates and/or probabilities. 

> Can you create a test sample, maybe also in TTree form, so that we can look at what it produces?

This module has the option to produce a TTree that I used for debugging. I can provide a copy of this TTree. It has an entry for each produced particle but maybe it's better to have an entry per stopped muon? If you would prefer this structure, then let me know :)

> In the meeting you presented this work, the request was made to expand this module to also produce primary events (conversion electrons, ...). Looking at the structure I don't see how that's natural, given that this module intrinsically generate multiple decay products. I can see how some infrastructure (GenProc, ...) might be shared with a module designed to produce primary events. That suggests those structs be split out of the module and put in their own files, and maybe made into classes.

I didn't realise this was a hard request. I was hoping to keep the scope of this PR to just adding a generator that reproduces all of our background frames in one. 

This generator could handle producing single particles like CeEndpoint, though. One would provide the process in the list of stopped muon processes with probability 1, and then a single particle would be produced in every event like we do now. However, something more complicated like back-to-back e+e- pairs would require a significant rethink.

Cheers
Andy

brownd1978 commented:
Hi Andy,

On Wed, Aug 26, 2020 at 4:04 PM Andrew Edmonds <notifications@github.com>
wrote:

> Hi Dave
>
> Can you please explain the normalization model of this module? Does each
> event corresponds to the decay or capture of a single muon, with the output
> containing all its products?
>
> Yes, each event corresponds to a single stopped muon that will either
> decay or capture. The user configures lists of processes that can occur for
> stopped muons, decaying muons, and captured muons. Each process has either
> a random chance of producing one particle (e.g. 79% for a 2p-1s X-ray), or
> a Poisson mean to produce multiple particles of the specified type (e.g.
> 1.2 for neutrons)
>
> The generator will go through the list of stopped muon processes in every
> event, but will only go through *either* the list of decay processes *or*
> the list of capture processes depending on the capture fraction.
>
> If so, is there any attempt to make sure energy and momentum and other
> quantum numbers are conserved?
>
> No, there isn't an attempt to conserve energy and momentum. I think the
> nuclear physics of the capture process complicates things too much.
>
> Or does each event corresponds to an incoherent sampling of independent
> muons, which just happen to stop at the same point, each of which may or
> may not produce a particle from the given process?
>
> I'm not sure. I don't quite understand the question, but I will think on
> it further.
>
Does each generator process always create a particle in every event? If so,
> how is the relative rate of (say) photon and neutron production maintained?
>
> No, each generator does not always produce a particle. It will depend on
> the emission rates and/or probabilities.
>
follow-up question; the mean number of neutrons and photons/capture is > 1,
can the generator produce >1 of these per event?


> Can you create a test sample, maybe also in TTree form, so that we can
> look at what it produces?
>
> This module has the option to produce a TTree that I used for debugging. I
> can provide a copy of this TTree. It has an entry for each produced
> particle but maybe it's better to have an entry per stopped muon? If you
> would prefer this structure, then let me know :)
>
An entry per stopped muon (1 event) would be best.  If that's difficult
technically just point us to the art output.


> In the meeting you presented this work, the request was made to expand
> this module to also produce primary events (conversion electrons, ...).
> Looking at the structure I don't see how that's natural, given that this
> module intrinsically generate multiple decay products. I can see how some
> infrastructure (GenProc, ...) might be shared with a module designed to
> produce primary events. That suggests those structs be split out of the
> module and put in their own files, and maybe made into classes.
>
> I didn't realise this was a hard request. I was hoping to keep the scope
> of this PR to just adding a generator that reproduces all of our background
> frames in one.
>
I agree it's outside the scope of this PR, but primary generation is a very
similar problem so I want to keep it in mind
as your use case is worked through.
 The 'hard requirement' is that we not duplicate code or configuration.  In
particular we don't want spectra, code for reading muon stop positions,
etc. implemented and configured in more than 1 place.


> This generator could handle producing single particles like CeEndpoint,
> though. One would provide the process in the list of stopped muon processes
> with probability 1, and then a single particle would be produced in every
> event like we do now. However, something more complicated like back-to-back
> e+e- pairs would require a significant rethink.
>
OK good.  Could you provide an example configuration of that for our
discussions?
thanks, Dave

Cheers
> Andy
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/231#issuecomment-681167955>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH57ZR5FAQX2BNQMQTV43SCWII5ANCNFSM4QGW5PWQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

gaponenko commented:
Hi,

On Wed, 26 Aug 2020, David Brown wrote:

> 
> > In the meeting you presented this work, the request was made to expand
> > this module to also produce primary events (conversion electrons, ...).
>
> I agree it's outside the scope of this PR, but primary generation is a very
> similar problem so I want to keep it in mind

I do not understand how handling CEs by the this generator can be
useful.  The main role of the new code as I see it is reducing the
number of mix-in background streams by replacing the separate proton,
neutron, dio, etc streams with a single "muon decay or capture"
*background* stream, with appropriate SM cross sections.

There is no SM cross sections for the signal, and producing a
background stream (with multiple events to mix in later) and signal
stream (with exactly one event to mix in later) in a single job does
not make sense.  Signal and background events should be produced by
separate jobs.  Then having the same art module to do both does not
improve anything, signal can as well be generated by a dedicated
module.

Andrei

brownd1978 commented:
Hi Andrei,

On Wed, Aug 26, 2020 at 4:58 PM Andrei Gaponenko <notifications@github.com>
wrote:

> Hi,
>
> On Wed, 26 Aug 2020, David Brown wrote:
>
> >
> > > In the meeting you presented this work, the request was made to expand
> > > this module to also produce primary events (conversion electrons, ...).
> >
> > I agree it's outside the scope of this PR, but primary generation is a
> very
> > similar problem so I want to keep it in mind
>
> I do not understand how handling CEs by the this generator can be
> useful. The main role of the new code as I see it is reducing the
> number of mix-in background streams by replacing the separate proton,
> neutron, dio, etc streams with a single "muon decay or capture"
> *background* stream, with appropriate SM cross sections.
>

Yes this is the main purpose.
However, we don't want to duplicate code for reading and processing mustops
(whatever format), which is also needed
for primary signals.  To have separate background stream and primary mustop
generators
requires making the mustop processing a utility used both modules.  That
could require more work, more maintenance
and less clarity than embedding both in a single module (as we do now).
I think it's a judgement call which approach is better, and I would leave
the decision to the code developer.


> There is no SM cross sections for the signal, and producing a
> background stream (with multiple events to mix in later) and signal
> stream (with exactly one event to mix in later) in a single job does
> not make sense. Signal and background events should be produced by
> separate jobs. Then having the same art module to do both does not
> improve anything, signal can as well be generated by a dedicated
> module.
>

That was not the use case I was thinking of, and I agree it makes no sense
to mix primary + backgrounds using this generator.
My use case was considering using this same module to produce a primary
particle from mustops.

cheers, Dave


>
> Andrei
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/231#issuecomment-681183364>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH575DJGQJ3Z7K3RAKXTLSCWOUDANCNFSM4QGW5PWQ>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

gaponenko commented:
Hi Dave,

On Wed, 26 Aug 2020, David Brown wrote:

> However, we don't want to duplicate code for reading and processing mustops
> (whatever format), which is also needed
> for primary signals.  

It is not duplicated.  There are 8 different generators that all share
sampling code from Mu2eUtilities/inc/RootTreeSampler.hh

> To have separate background stream and primary mustop
> generators
> requires making the mustop processing a utility used both modules.  That
> could require more work, more maintenance
> and less clarity than embedding both in a single module (as we do now).

The module part is easy; the amount of boilerplate code is pretty
trivial.  An advantage of having individual modules is that one can do
away with handling a configuration that specifies what process to
generate.  Instead simply point to the desired module_type in fhicl.
This is a way to reduce the amount of code to write and maintain - by
delegating the work to the art infrastructure.

We already use standalone generators for other high-energy track
processes (RMC, RPC).  I initially wrote StoppedParticleReactionGun as
a single module with multiple processes, but it was before the
RootTreeSampler code was factored out.  I would have used standalone
modules for each process if I were to re-do it today.

> I think it's a judgement call which approach is better, and I would leave
> the decision to the code developer.
> ...
> My use case was considering using this same module to produce a primary
> particle from mustops.

In the MuStopProductsGun case one has to already handle an elaborate
configuration with mulitple processes, so the incremental cost of
adding CE may be small.   May be it is reasonable here.

Andrei

gaponenko reviewed (changes requested):
Hi Andy,

Simulating all of our "background hits" processes in a single job is a
good idea; it can significantly simplify production of mixed datasets.
I see two conceptual ways of producing the background stream:

a) Follow the physics at the microscopic level: define a set of
   exclusive physics channels with specified branching fractions and
   make each muon go through exactly one of them. (If it decays it
   does not get captured.  Then one can have a list of "real" physics
   channels like n+gamma, 2n, n+p etc. emissions for the case of
   capture.  Or some simplified process list.)

b) Follow the physics on average.  This can be handled by defining a
   set of channels with probabilities of the process attached to each
   channel.  Some muons will both decay and capture, some will just
   disappear.  If we do not care about correlations of daughter
   products this approach should be sufficient.

It looks like the present code follows a hybrid approach, with decay
and capture being exclusive, but then "everything can happen at once"
within each category.  This is fine as well.

While I like the idea, there are things in the present code that I
think should be changed.

* Physics

   This PR makes an attempt to simulate the physics of the atomic
   capture in addition to muon decay and nuclear capture.  This is a
   completely new development, which needs to be at least discussed
   specifically and not in the context of re-factoring the exactly the
   same physics as we used in the past into a single art module.  That
   would affect other parts of the simulation.  Notably, the job that
   records muon stops in the target already simulates the atomic
   cascade using GEANT4; this would need to be disabled to avoid
   double-counting.  But I am not aware of anyone demonstrating the
   need to improve on G4 atomic cascade simulation for the backgrounds
   hit stream, so the motivation to do it "by hand" instead of using
   out of the box G4 is not clear.  On the surface G4 has a more
   detailed simulation of the process, which includes e.g. the
   emission of Auger electrons, which is lacking in this PR.

   I request that all atomic cascade stuff to be removed from this PR.
   (Not just from the module but e.g. in GlobalConstantsService).
   

* Configuration

  The GenPhysConfig structure is a union of all fields used by any of
  the generators.  Then the top level configuration includes sequences
  of those, yielding Cartesian products of (set of
  generators)x(configs of all generators).  That set is so "rich" that
  most defined fields are not used, defeating the purpose of FHICL
  validation.  And it still does not have enough information for
  potential future channels. (E.g. correlated emission of proton and
  neutron.)

  I suggest to rework the code to use art plugin tools. Couple
  possible ways to do it:
  
  a) Explicitly define a configuration struct for each
     decay/capture channel, and instead of e.g.
     
     Sequence<Table<GenPhysConfig> > captureProducts{...};

     at the top level, list

     Table<CaptureProtonGunConfig> proton{...};
     Table<CaptureNeutronGunConfig> neutron{...};
     ...

   b) Use art plugin tools. Each physics channel can be handled by its
      tool, with individual tool configuration completely
      FHICL-validated.  See the discussion on the art-users list.

      https://listserv.fnal.gov/scripts/wa.exe?A0=ART-USERS

* BinnedSpectrum use.  Perhaps the configuration issue is at least
  partly explained by the reliance on the BinnedSpectrum class.  That
  class started as a simple utility, and morphed into a monster that
  knows about radiative corrections to CE electrons, the RMC process,
  Czarnecki spectrum, and what not...  Of course its config can not
  make sense.  I suggest that new code should not depend on
  BinnedSpectrum, so we can eventually retire it.  That will require
  taking pieces of BinnedSpectrum code and placing them in new helper
  classes with a better defined overall structure.  At least those that
  you immediately need.  This is a big request; if it's too big for
  this PR we should create an issue for that.

* Remove histogramming to simplify the code.  The output of the code
  can be monitored by an analyzer module running on
  GenParticleCollection when desired; there is no reason to put both
  "generator" and "analyzer" into a single module here.

* Fix whitespace, e.g. hard tabs and trailing whitespace in FHICL.

gaponenko commented:
A correction:

>   I suggest to rework the code to use art plugin tools. 

The approach (a) there has nothing to do with the tools.

Andrei

AndrewEdmonds11 commented:
Hi Andrei
> 
> ...
> It looks like the present code follows a hybrid approach, with decay
> and capture being exclusive, but then "everything can happen at once"
> within each category. This is fine as well.
> ...

Yeh, this hybrid approach arises because we don't know the correlations of the capture products. I don't see us having this information in the near future.

> ...
>       I request that all atomic cascade stuff to be removed from this PR.
>       (Not just from the module but e.g. in GlobalConstantsService).
> ...

I'm happy to do this. It can be added at a later date if we discover we need it.

> 
> ...
>       I suggest to rework the code to use art plugin tools. Couple
>       possible ways to do it:
>       a) Explicitly define a configuration struct for each
>       decay/capture channel, and instead of e.g.
>       Sequence<Table > captureProducts{...};
>       at the top level, list
>       Table proton{...};
>       Table neutron{...};
>       ...
>       b) Use art plugin tools. Each physics channel can be handled by its
>       tool, with individual tool configuration completely
>       FHICL-validated. 
> ...

Thanks, I will look into these options. My gut reaction is that this might remove some flexibility that I was hoping to keep (in terms of particle and spectrum choice) but I will look into it further.

>     * BinnedSpectrum use.  
> ...
>   This is a big request; if it's too big for
>       this PR we should create an issue for that.
> ...

Yes, I think this a big job and beyond the scope of this PR. I would rather the new generator uses BinnedSpectrum as-is, which has been validated, than try to tease out parts into utilities that will need revalidating.

> 
>     * Remove histogramming to simplify the code.  The output of the code
>       can be monitored by an analyzer module running on
>       GenParticleCollection when desired; there is no reason to put both
>       "generator" and "analyzer" into a single module here.
> 
>     * Fix whitespace, e.g. hard tabs and trailing whitespace in FHICL.

I'm happy to do both of these.

ryuwd commented:
@FNALbuild run code checks

FNALbuild commented:
:hourglass: The following tests have been triggered for 3443e1dec64e774c8a4e9762f9c53844fbe6a09c: code checks

FNALbuild commented:
:x: Trailing whitespace characters were found at 3443e1dec64e774c8a4e9762f9c53844fbe6a09c on files you changed.

You can review the generated patch [here](https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-code-checks//108/artifact/codechecks-pr231-3443e1dec64e774c8a4e9762f9c53844fbe6a09c.patch).

If it is convenient, you may remove the trailing whitespace like this:
```
curl https://buildmaster.fnal.gov/buildmaster/job/GitHubPRTests/job/mu2e-offline-code-checks//108/artifact/codechecks-pr231-3443e1dec64e774c8a4e9762f9c53844fbe6a09c.patch | git apply -v --index
git commit -am "Remove trailing whitespace found at 3443e1dec64e774c8a4e9762f9c53844fbe6a09c" && git push
```

gaponenko commented:
Hi Andy,

> >       ...
> >       b) Use art plugin tools. Each physics channel can be handled by its
> >       tool, with individual tool configuration completely
> >       FHICL-validated. 
> > ...
> 
> Thanks, I will look into these options. My gut reaction is that this might remove some flexibility that I was hoping to keep (in terms of particle and spectrum choice) but I will look into it further.

What I have in mind is defining an interface like

class IMuStopParticleGenerator {
public:
  virtual void generate(GenParticleCollection *out, const StoppedParticleF& stop) = 0;
  virtual ~IMuStopParticleGenerator() = default;
};

where the generate() function takes muon stop position and time,
creates the desired particle(s), and adds them to a pre-defined
collection.  So that several generators can be called in sequence on
one event, as you do in the present code.  Also, the interface allows
a generator to produce multiple correlated particles if we come to
this.

Art class tools
https://cdcvs.fnal.gov/redmine/projects/art/wiki/Guide_to_writing_and_using_tools
can override the generate() method and take care of individual
processes, while being completely decoupled from each other.
I do not see where flexibility is missing.

Best regards,
Andrei

AndrewEdmonds11 commented:
This is now ready for people to have a final look through. Here is the run-down:

 - added a new generator where each event corresponds to a stopped muon
 - the new generator uses new plugin tools to generate each process
 - I have added constants to the GlobalConstantsService to handle particle emission rates
 - this is not a breaking change

Once I get the go ahead, I will close this PR, merge it with PRs #234 and #243 and then open a new pull request
